### PR TITLE
Switch RBAC subject apiVersion to apiGroup in v1beta1

### PR DESCRIPTION
**Pull request by liggitt**
Referencing a subject from an RBAC role binding, the API group and kind of the subject is needed to fully-qualify the reference.

The version is not, and adds complexity around re-writing the reference when returning the binding from different versions of the API, and when reconciling subjects.

This PR:
* v1beta1: change the subject `apiVersion` field to `apiGroup` (to match roleRef)
* v1alpha1: convert apiVersion to apiGroup for backwards compatibility
* all versions: add defaulting for the three allowed subject kinds
* all versions: add validation to the field so we can count on the data in etcd being good until we decide to relax the apiGroup restriction

```release-note
RBAC `v1beta1` RoleBinding/ClusterRoleBinding subjects changed `apiVersion` to `apiGroup` to fully-qualify a subject. ServiceAccount subjects default to an apiGroup of `""`, User and Group subjects default to an apiGroup of `"rbac.authorization.k8s.io"`.
```

@deads2k @kubernetes/sig-auth-api-reviews @kubernetes/sig-auth-pr-reviews

### CODE REVIEW

File: pkg/apis/rbac/types.go [hunk 1]
```diff
@@ -63,9 +63,9 @@ type Subject struct {
 	// Kind of object being referenced. Values defined by this API group are "User", "Group", and "ServiceAccount".
 	// If the Authorizer does not recognized the kind value, the Authorizer should report an error.
 	Kind string
-	// APIVersion holds the API group and version of the referenced object. For non-object references such as "Group" and "User" this is
-	// expected to be API version of this API group. For example, "rbac/v1alpha1".
-	APIVersion string
+	// APIGroup holds the API group of the referenced object. For non-object references such as "Group" and "User" this is
+	// expected to be this API group. For example, "rbac.authorization.k8s.io".
+	APIGroup string
```
Comment (deads2k) [inline pkg/apis/rbac/types.go]: I'm in favor, even though most places stuff groups into the version field.

Comment (ericchiang) [inline pkg/apis/rbac/types.go]: Can we articulate that setting this value isn't required for "normal" uses and defaults to a reasonable value? We've had users consistently be confused by this field.

**Comment by liggitt** [comment]
@ericchiang this better?

```
// APIGroup holds the API group of the referenced subject.
// Defaults to "" for ServiceAccount subjects.
// Defaults to "rbac.authorization.k8s.io" for User and Group subjects.
```

also added validation so that if you fat-finger some other value, it won't get persisted and it'll tell you what it expected

**Comment by liggitt** [comment]
cc @kubernetes/api-approvers for approval

**Comment by ericchiang** [comment]
@liggitt looks great.

File: pkg/apis/rbac/validation/validation.go [hunk 2]
```diff
@@ -201,6 +201,9 @@ func validateRoleBindingSubject(subject rbac.Subject, isNamespaced bool, fldPath
 				allErrs = append(allErrs, field.Invalid(fldPath.Child("name"), subject.Name, msg))
 			}
 		}
+		if subject.APIGroup != "" {
```
Comment (smarterclayton) [inline pkg/apis/rbac/validation/validation.go]: nit: use len() for consistency with the rest of this method

Comment (liggitt) [inline pkg/apis/rbac/validation/validation.go]: done

File: pkg/apis/rbac/v1alpha1/zz_generated.conversion.go [hunk 3]
```diff
@@ -410,20 +410,16 @@ func Convert_rbac_RoleRef_To_v1alpha1_RoleRef(in *rbac.RoleRef, out *RoleRef, s
 
 func autoConvert_v1alpha1_Subject_To_rbac_Subject(in *Subject, out *rbac.Subject, s conversion.Scope) error {
 	out.Kind = in.Kind
-	out.APIVersion = in.APIVersion
+	// WARNING: in.APIVersion requires manual conversion: does not exist in peer-type
```
Comment (smarterclayton) [inline pkg/apis/rbac/v1alpha1/zz_generated.conversion.go]: Set the conversion=false tag to avoid this warning

**Review by smarterclayton — CHANGES_REQUESTED** [review]
Code looks fine, API looks fine.  One comment

File: pkg/apis/rbac/v1alpha1/zz_generated.conversion.go (hunk 3 above)
Comment (liggitt) [inline pkg/apis/rbac/v1alpha1/zz_generated.conversion.go]: fixed

**Comment by smarterclayton** [comment]
Any other comments folks want to add?

**Review by smarterclayton — APPROVED** [review]

**Comment by liggitt** [comment]
tagging per https://github.com/kubernetes/kubernetes/pull/41184#pullrequestreview-21374999

**Pull request merged**